### PR TITLE
Changing how the function (cutTo)firstSignificantSubdomain detect the tld (proposal)

### PR DESCRIPTION
**Pull request by YiuRULE**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description (up to few sentences):

As suggested in a comment, now for detect what is the second tld. It's using a perfect hash table generated by gperf with a list generated from this file : https://publicsuffix.org/list/public_suffix_list.dat, who can give a better accuracy for some result (for example, with url who contains the domain `ac.uk`).

Detailed description (optional):

So here a potential proposal. Would just like to tell few points, it would add gperf as a dependency for the project as the code is generated on the compilation phase with `ninja-build`.
Also I wasnt' really sure for to things, with two domain from the test, one who have co.yandex and org.рф, I wasn't able to find this domain both on the list and also doing a quick research from Google, so let me know if I should add them on the list too.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
> Also I wasnt' really sure for to things, with two domain from the test, one who have co.yandex and org.рф, I wasn't able to find this domain both on the list and also doing a quick research from Google, so let me know if I should add them on the list too.

These domains fit to the older logic of this function, but they don't exist actually.
Modified test is Ok.

**Comment by alexey-milovidov** [comment]
@YiuRULE You should also add dependency to Dockerfiles in `docker` directory, whenever appropriate.

**Comment by YiuRULE** [comment]
@alexey-milovidov Ok thanks for your help, I added the dependency in the `builder` folder, I think it should be ok now

**Comment by alexey-milovidov** [comment]
It's not enough:
https://clickhouse-builds.s3.yandex.net/5030/5c67ecd46d803f97908563b27fc9cf9a95f7b393/build_log_415594408_1555584319.txt

As far as I remember, you should add to `packager` Dockerfiles.

**Comment by YiuRULE** [comment]
Thanks, I add `gperf` dependency both on `binary` and `deb` folder.

Otherwise it seem that we still have a problem with the build, it seem that CI fetch the docker image from the one in the [docker hub]([https://hub.docker.com/r/yandex/clickhouse-deb-builder]() right?

Otherwise, I tried running locally the script `packager` with the command line `./packager  --package-type deb --output-dir build` and seem the build went fine :/

**Comment by alesapin** [comment]
@YiuRULE unfortunately in our CI system Docker images have to be pushed by hands. I pushed images containing gperf. Can you merge with master or do any insignificant commit?

**Comment by alexey-milovidov** [comment]
@alesapin I've made a commit.

**Comment by alexey-milovidov** [comment]
Need to use more loose compiler warning flags for compiling auto-generated file `tldLookup.cpp`

**Comment by alexey-milovidov** [comment]
Performance tests are Ok.

**Comment by alexey-milovidov** [comment]
> We don't have anymore build problem using Clang, does this fix seem ok ?

Yes.

File: dbms/src/Functions/CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,12 @@ include(${ClickHouse_SOURCE_DIR}/cmake/dbms_glob_sources.cmake)
 add_headers_and_sources(clickhouse_functions ./GatherUtils)
 add_headers_and_sources(clickhouse_functions .)
 
+add_custom_command(
+	OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/tldLookup.cpp
+	COMMAND ${GPERF} ${CMAKE_CURRENT_SOURCE_DIR}/gperf/tldLookup.gperf --output-file=${CMAKE_CURRENT_SOURCE_DIR}/tldLookup.cpp
```
Comment (proller) [inline dbms/src/Functions/CMakeLists.txt]: Why use CMAKE_CURRENT_SOURCE_DIR?
Much better is CMAKE_CURRENT_BINARY_DIR for auto-generated files.

File: docs/en/development/build.md [hunk 2]
```diff
@@ -70,6 +70,11 @@ export CXX=g++-7
 sudo apt-get install libicu-dev libreadline-dev
 ```
 
+## Install Additional dependencies
+```bash
+sudo apt-get install gperf
```
Comment (proller) [inline docs/en/development/build.md]: Just add gperf to previous `sudo apt-get install`

File: dbms/src/Functions/CMakeLists.txt [hunk 3]
```diff
@@ -3,6 +3,12 @@ include(${ClickHouse_SOURCE_DIR}/cmake/dbms_glob_sources.cmake)
 add_headers_and_sources(clickhouse_functions ./GatherUtils)
 add_headers_and_sources(clickhouse_functions .)
 
+add_custom_command(
```
Comment (proller) [inline dbms/src/Functions/CMakeLists.txt]: if (USE_GPERF)

File: dbms/src/Functions/firstSignificantSubdomain.h [hunk 4]
```diff
@@ -2,7 +2,7 @@
```
Comment (proller) [inline dbms/src/Functions/firstSignificantSubdomain.h]: #include <Common/config.h>
#if USE_GPERF

**Comment by YiuRULE** [comment]
Thanks, I applied the change.

It seem we got a weird bug with the CI process when tried to build using g++ (both on this branch, and the latest master version).

```
[7/87] cd /home/guillaume/ClickHouse &...d/dbms/programs/clang/headers/19.7.1.1
Use one of the following commands:
clickhouse local [args] 
clickhouse client [args] 
clickhouse benchmark [args] 
clickhouse server [args] 
clickhouse performance-test [args] 
clickhouse extract-from-config [args] 
clickhouse compressor [args] 
clickhouse format [args] 
clickhouse copier [args] 
clickhouse obfuscator [args] 
```

Should I do something in particular on this branch or should be ok ?

**Comment by YiuRULE** [comment]
Any news concerning this PR ? saw some commits who were pushed :)

**Comment by alexey-milovidov** [comment]
Merged. Sorry for the delay.

**Comment by proller** [comment]
@YiuRULE Can you show how you exactly generate list from  https://publicsuffix.org/list/public_suffix_list.dat ?

**Comment by YiuRULE** [comment]
@proller I create a simple python script who set the header of the gperf file

```
%language=C++
%define lookup-function-name is_valid
%define class-name tldLookupHash
%readonly-tables
%includes
%compare-strncmp
%{
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wimplicit-fallthrough"
#pragma GCC diagnostic ignored "-Wzero-as-null-pointer-constant"
#pragma GCC diagnostic ignored "-Wunused-macros"
%}
# List generated using https://publicsuffix.org/list/public_suffix_list.dat
%%
```

then the python script take the file as a parameter, remove every comment and empty line

then, it set this as a footer
```
%%
```

**Pull request merged**